### PR TITLE
3730-1

### PR DESCRIPTION
**Pull request by nikitathebest**
## Номер
3730

## Номер задания
1

## Комментарии
Run bundle install

### CODE REVIEW

File: 3730/1/gem_info.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
+require 'rest-client'
+require 'nokogiri'
+require 'open-uri'
+require 'terminal-table'
+
+# :reek:TooManyInstanceVariables
```
Comment (longbeardmen) [inline 3730/1/gem_info.rb]: this should be fixed

File: 3730/1/gem_info.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    @used_by = used_by
+    @watched_by = watched_by
+    @stars = stars
+    @forks = forks
+    @contributors = contributors
+    @issues = issues
+  end
+
+  def calculate_score
+    @score = @used_by.to_i / 5 + @watched_by.to_i / 2 + @stars.to_i +
```
Comment (longbeardmen) [inline 3730/1/gem_info.rb]: what're the magic numbers?

File: 3730/1/gem_info.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  def parse_github
+    @parse_github ||= Nokogiri::HTML(::Kernel.open(@github_url))
+  end
+
+  def parse_github_dependents
+    Nokogiri.HTML(::Kernel.open("#{@github_url}/network/dependents"))
+  end
+
+  def watched_by
+    parse_github.css("a[class='social-count']")[0].text.tr('^0-9', '')
```
Comment (longbeardmen) [inline 3730/1/gem_info.rb]: `.tr('^0-9', '')` called in couple methods
it could be called later

File: 3730/1/gem_info.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def stars
+    parse_github.css("a[class='social-count js-social-count']").text.tr('^0-9', '')
```
Comment (longbeardmen) [inline 3730/1/gem_info.rb]: this and the following methods do pretty same job:
`parse_github.css(css_path)[some index].text.tr('^0-9', '')`
it should be automated

File: 3730/1/repository.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    true
+  rescue RestClient::NotFound
+    puts "No such gem #{name}"
+    false
+  end
+
+  def self.find_github_url(name)
+    url = "https://api.github.com/search/repositories?q=#{name}&per_page=1"
+    response = RestClient.get(url)
+    JSON.parse(response)['items'][0]['html_url']
```
Comment (longbeardmen) [inline 3730/1/repository.rb]: you do
```
url = ...
response = RestClient.get(url)
```
twice.
It should be move to separate method

File: 3730/1/top_gems.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
+require 'yaml'
+require 'terminal-table'
+require 'optparse'
+require_relative './repository'
```
Comment (longbeardmen) [inline 3730/1/top_gems.rb]: `./` is redundant

File: 3730/1/top_gems.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  table.title = 'Gems statistics'
+  table.headings = ['Name', 'Used by', 'Watched by', 'Star', 'Forks', 'Contributors', 'Issues']
+  puts table
+end
+
+def sorted_gems(gems)
+  gems.sort! { |first, second| second.score <=> first.score }
+
+  number = if @command_line[:number]
+             @command_line[:number].to_i
```
Comment (longbeardmen) [inline 3730/1/top_gems.rb]: it should fit in ternary operator

File: 3730/1/top_gems.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  gems[0...number].each { |gem| row.add_row gem.output }
+end
+
+def print_table(gems, number)
+  table = Terminal::Table.new do |row|
+    add_row_to_table(gems, row, number)
+  end
+
+  table.title = 'Gems statistics'
+  table.headings = ['Name', 'Used by', 'Watched by', 'Star', 'Forks', 'Contributors', 'Issues']
```
Comment (longbeardmen) [inline 3730/1/top_gems.rb]: 1 should be a constant 
2 %w

File: 3730/1/top_gems.rb [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+def find_gem(name)
+  return unless Repository.a_gem?(name)
+  url = Repository.find_github_url(name)
+  gem = GemInfo.new(name, url)
+  gem.calculate_score
+  @gems << gem
+end
+
+# :reek:TooManyStatements
```
Comment (longbeardmen) [inline 3730/1/top_gems.rb]: should be fixed

File: 3730/1/top_gems.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+@command_line = {}
+
+OptionParser.new do |option|
+  option.on('--top NUMBER') { |i| @command_line[:number] = i }
+  option.on('--name NAME') { |i| @command_line[:name] = i }
+  option.on('--file FILE') { |i| @command_line[:file_name] = i }
+end.parse!
+
+filename = if @command_line[:file_name]
```
Comment (longbeardmen) [inline 3730/1/top_gems.rb]: filename = @command_line[:file_name] || 'gems.yml'

File: 3730/1/repository.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+require 'rest-client'
+require 'json'
+
+class Repository
+  def self.a_gem?(name)
+    url = "https://rubygems.org/api/v1/gems/#{name}.json"
+    RestClient.get(url)
+    true
```
Comment (longbeardmen) [inline 3730/1/repository.rb]: I would say it's redundant HTTP call

**Comment by nikitathebest** [comment]
@longbeardmen Issues fixed

File: 3730/1/gem_info.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+  def initialize(name, github_url)
+    @github_url = github_url
+    @gem_info = { name: name, used_by: used_by, watched_by: watched_by,
+                  stars: stars, forks: forks, contributors: contributors, issues: issues }
+  end
+
+  def calculate_score
+    @score = 0
+    gem_info.each_value { |value| @score += value.to_i }
```
Comment (longbeardmen) [inline 3730/1/gem_info.rb]: @score = gem_info.values.inject(0) { |sum, value| sum += value.to_i }

File: 3730/1/table.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+require 'terminal-table'
+
+class Table
+  HEADINGS = %w[Name Used\ by Watched\ by Star Forks Contributors Issues].freeze
+
+  def self.add_row_to_table(gems, row, number)
+    gems[0...number].each { |gem| row.add_row gem.output }
```
Comment (longbeardmen) [inline 3730/1/table.rb]: gems.first(number).each { ... }

**Pull request merged**